dotnet/src/Connectors/Connectors.Memory.Pinecone/PineconeVectorStoreRecordMapper.cs

```diff
@@ -0,0 +1,194 @@
+
+        this._keyPropertyInfo = properties.keyProperty;
+        this._payloadPropertiesInfo = properties.dataProperties;
+        this._vectorPropertyInfo = properties.vectorProperties.First();
+        this._storagePropertyNames = VectorStoreRecordPropertyReader.BuildPropertyNameToStorageNameMap(properties, this._options.VectorStoreRecordDefinition);
+    }
+
+    /// <inheritdoc />
+    public Vector MapFromDataToStorageModel(TRecord dataModel)
+    {
+        var keyObject = this._keyPropertyInfo.GetValue(dataModel);
+        if (keyObject is null)
+        {
+            throw new VectorStoreRecordMappingException($"Key property {this._keyPropertyInfo.Name} on provided record of type {typeof(TRecord).FullName} may not be null.");
+        }
+
+        var metadata = new MetadataMap();
+        foreach (var payloadPropertyInfo in this._payloadPropertiesInfo)
+        {
+            var propertyName = this._storagePropertyNames[payloadPropertyInfo.Name];
+            var propertyValue = payloadPropertyInfo.GetValue(dataModel);
+            if (propertyValue != null)
+            {
+                metadata[propertyName] = ConvertToMetadataValue(propertyValue);
+            }
+        }
+
+        var valuesObject = this._vectorPropertyInfo.GetValue(dataModel);
+        if (valuesObject is not ReadOnlyMemory<float> values)
+        {
+            throw new VectorStoreRecordMappingException($"Vector property {this._vectorPropertyInfo.Name} on provided record of type {typeof(TRecord).FullName} may not be null.");
+        }
+
+        // TODO: what about sparse values?
+        var result = new Vector
+        {
+            Id = (string)keyObject,
+            Values = values.ToArray(),
+            Metadata = metadata,
+            SparseValues = null
+        };
+
+        return result;
+    }
+
+    /// <inheritdoc />
+    public TRecord MapFromStorageToDataModel(Vector storageModel, StorageToDataModelMapperOptions options)
+    {
+        var keyPropertyName = this._storagePropertyNames[this._keyPropertyInfo.Name];
+
+        var outputJsonObject = new JsonObject
+        {
+            { this._keyPropertyInfo.Name, JsonValue.Create(storageModel.Id) },
+        };
+
+        if (options?.IncludeVectors is true)
+        {
+            var propertyName = this._storagePropertyNames[this._vectorPropertyInfo.Name];
+            outputJsonObject.Add(this._vectorPropertyInfo.Name, new JsonArray(storageModel.Values.Select(x => JsonValue.Create(x)).ToArray()));
+        }
+
+        if (storageModel.Metadata != null)
+        {
+            foreach (var payloadProperty in this._payloadPropertiesInfo)
+            {
+                var propertyName = this._storagePropertyNames[payloadProperty.Name];
+                if (storageModel.Metadata.TryGetValue(propertyName, out var value))
+                {
+                    outputJsonObject.Add(payloadProperty.Name, ConvertFromMetadataValueToJsonNode(value));
+                }
```

Comment:
I found a small issue with this approach that I had to fix in the QdrantVectorStoreRecordMapper implementation.
If someone decorated their model with JsonPropertyName attributes, then the property name that Deserialize expects will be different to the name of the model property.
I created a helper method `VectorStoreRecordPropertyReader.GetJsonPropertyName` that makes it easy to find the correct json property name to use.
So it means that we would continue to read values from the storage model using the storage property names, but add these values to the json object using the json property names.

To test it out I added a `JsonPropertyName` attribute with a very different name to one of my test model properties.  See the `DataInt` property on the `MultiPropsModel<TKey>` class in `QdrantVectorStoreRecordMapperTests`.